src/main/scala/monocly/Optic.scala

```diff
@@ -0,0 +1,94 @@
+package monocly
+
+import monocly.impl._
+import monocly.functions.{Each, Index}
+import monocly.internal.{NonEmptyList, Applicative, Monoid}
+
+
+sealed trait OpticCan
+
+//        GetMany
+//          /\
+//         /  \
+// GetOption  GetOneOrMore
+//         \  /
+//          \/
+//        GetOne
+trait GetMany extends OpticCan
+trait GetOption extends GetMany
+trait GetOneOrMore extends GetMany
+trait GetOne extends GetOption with GetOneOrMore 
+
+//    Modify
+//      ^
+//      |
+//  ReverseGet
+trait Modify extends OpticCan
+trait ReverseGet extends Modify
+
+type Optic[+ThisCan <: OpticCan, S, A] = POptic[ThisCan, S, S, A, A]
+
+
+final class POptic[+ThisCan <: OpticCan, -S, +T, +A, -B] private[monocly](
+    protected[monocly] val getter: GetterImpl[ThisCan, S, A],
+    protected[monocly] val setter: SetterImpl[ThisCan, S, T, A, B]):
+
+  def andThen[ThatCan <: OpticCan, BothCan >: (ThisCan | ThatCan) <: OpticCan, C, D](o: POptic[ThatCan, A, B, C, D]): POptic[BothCan, S, T, C, D] = 
+    POptic(getter.andThen(o.getter), setter.andThen(o.setter))
+
+end POptic
+
+
+extension [ThisCan <: OpticCan, S, A] (optic: Optic[ThisCan, S, A])
+  def each[C](using evEach: Each[A, C]): Optic[ThisCan | (GetMany & Modify), S, C] = 
+    optic.andThen(evEach.each)
+
+extension [S, T, A, B] (optic: POptic[GetMany, S, T, A, B])
+  inline def foldMap[M: Monoid](f: A => M)(s: S): M = ???
+
+extension [S, T, A, B] (optic: POptic[GetOne, S, T, A, B])
+  inline def get: S => A = optic.getter.get
```

Comment:
it is a detail but do you think we could/should move these methods to the POptic class? e.g.
```scala
final class POptic[+ThisCan <: OpticCan, -S, +T, +A, -B]  ... {
   inline def get(using ev: ThisCan <: GetOne): S => A = optic.getter.get
}
```

Reply:
I agree, the code is cleaner this way, but if we use class methods, we will avoid issues with other libraries using the same extension method name (e.g. `index` with cats)